### PR TITLE
Add support for IFileOperation interface

### LINKED ISSUE #1: Use the newer IFileOperation interface
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/recycle-bin/issues/1)

From the [SHFileOperation](https://msdn.microsoft.com/en-us/library/windows/desktop/bb762164%28v=vs.85%29.aspx) docs:

> This function has been replaced in Windows Vista by IFileOperation.
## 

I wanted to use [IFileOperation](https://msdn.microsoft.com/en-us/library/windows/desktop/bb775771%28v=vs.85%29.aspx) initially, but couldn't find any good examples of using it in C. Most used it in C++.

Should use the [`IFileOperation::DeleteItems`](https://msdn.microsoft.com/en-us/library/windows/desktop/bb775767%28v=vs.85%29.aspx) and [`IFileOperation::SetOperationFlags(FOFX_ADDUNDORECORD | FOFX_RECYCLEONDELETE | FOF_NOERRORUI | FOF_SILENT | FOFX_EARLYFAILURE)`](https://msdn.microsoft.com/en-us/library/windows/desktop/bb775799%28v=vs.85%29.aspx) methods.
## 

Switching to then newer API might add support for [long paths](https://msdn.microsoft.com/en-us/library/aa365247%28VS.85%29.aspx), not really sure.
See: https://github.com/sindresorhus/trash/issues/16.
## 

We can still keep the `SHFileOperation` around for support for Vista and older, as the `FOFX_RECYCLEONDELETE` flag requires Windows 8.
## 

**Help wanted** - My C skills are very much at a noobish level, so any help would be much appriciated :)

Happy to use C++ if that's what it takes. I'm just trying to keep it simple with C.

> There is a $100.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/recycle-bin/issues/1).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/recycle-bin/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by medusalix**
I've managed to implement support for the `IFileOperation` interface in pure C 😃. Right now, the program falls back to the old `SHFileOperation` function for versions lower than Windows 8, but I think we should remove support for the old API altogether, as recycle on delete is the default behavior with `IFileOperation` for Windows 7 and Vista, according to [Electron](https://github.com/electron/electron/blob/master/atom/common/platform_util_win.cc#L345). Dealing with `COM interfaces` in C/C++ is a little bit ugly in general, so I've added some [macros](https://github.com/medusalix/recycle-bin/blob/e495371c08d7921e126ab19e8c49742413e40599/recycle-bin.c#L10-L16) to help with error checking. Not sure if that's the best way to do so, open for improvements.

I added some comments for stuff that requires a little explanation to understand.

Resolves #1

### CODE REVIEW

File: build.bat [hunk 1]
```diff
@@ -2,4 +2,4 @@
 
 set filename=recycle-bin
 
-gcc "%filename%".c -municode -O2 -s -o "%filename%".exe -std=c99
+gcc "%filename%".c -municode -O2 -lole32 -s -o "%filename%".exe -std=c99
```
Comment (medusalix) [inline build.bat]: Linking against `Ole32` is necessary for `COM` to work.

File: recycle-bin.c [hunk 2]
```diff
@@ -1,38 +1,34 @@
+#define _WIN32_IE _WIN32_IE_IE70
```
Comment (medusalix) [inline recycle-bin.c]: This `define` is required for `shlobj.h` to include the `IFileOperation` interface.

File: recycle-bin.c [hunk 3]
```diff
@@ -53,3 +49,69 @@ int wmain(int argc, wchar_t **argv) {
...
+		FOFX_ADDUNDORECORD |
+		FOFX_RECYCLEONDELETE |
+		FOF_NOERRORUI |
+		FOF_SILENT |
+		FOFX_EARLYFAILURE
+	));
+
+	PCIDLIST_ABSOLUTE list[count - 1];
+
+	for (int i = 1; i < count; i++) {
```
Comment (medusalix) [inline recycle-bin.c]: I found out that `IFileOperation` refuses to work with paths that are separated by forward slashes. This loop converts all paths from forward slashes to backward slashes.

**Comment by sindresorhus** [comment]
> but I think we should remove support for the old API altogether

👍

File: recycle-bin.c (hunk 3 above)
Comment (sindresorhus) [reply recycle-bin.c]: Is there really no system API to do this?

https://docs.microsoft.com/en-us/windows/desktop/FileIO/naming-a-file

Couldn't we make the paths "long paths" (`\\?\`-prefixed)? Those support forward slashes and they also don't have length restrictions.

Comment (sindresorhus) [reply recycle-bin.c]: Also note:

> Starting in Windows 10, version 1607, MAX_PATH limitations have been removed from common Win32 file and directory functions. However, you must opt-in to the new behavior.
> You can also enable the new long path behavior per app via the manifest:

We should probably do that.

**Comment by sindresorhus** [comment]
Can you add a test for a long file path (more than 255 characters)? So we can ensure that works now.

File: recycle-bin.c (hunk 3 above)
Comment (medusalix) [reply recycle-bin.c]: I tried to prefix the paths with `\\?\` before passing them to `IFileOperation`, but apparently that doesn't work at all (not even for short paths). It seems like the Windows Shell API doesn't support long paths at all ([MSDN](https://social.msdn.microsoft.com/Forums/vstudio/en-US/42353ae3-1db9-45a2-a30f-99c184355b71/do-shell-objects-accept-long-path-names#6e53c00c-2d30-4692-a28e-253bbd6b04dd), [Stack Overflow](https://stackoverflow.com/questions/9980943/bypassing-max-path-limitation-for-itemidlist#comment12757715_9981012)). Although in theory it shouldn't accept paths longer than `MAX_PATH`, it still managed to delete a path with a length of over 400 characters in my Windows 7 VM 😄.

**Comment by medusalix** [comment]
How do you want me to add a test? Could you please elaborate?

File: recycle-bin.c (hunk 3 above)
Comment (medusalix) [reply recycle-bin.c]: > Starting in Windows 10, version 1607, MAX_PATH limitations have been removed from common Win32 file and directory functions. However, you must opt-in to the new behavior.
> You can also enable the new long path behavior per app via the manifest:

> These are the directory management functions that no longer have MAX_PATH restrictions if you opt-in to long path behavior: CreateDirectoryW, CreateDirectoryExW, ...

>These are the file management functions that no longer have MAX_PATH restrictions if you opt-in to long path behavior: CopyFileW, CopyFile2, ...

I think this only applies to the lower `Win32 File System` functions. The `Windows Shell API` is still [restricted](https://stackoverflow.com/questions/23976112/do-windows-shell-apis-work-with-long-unicode-paths).

**Comment by sindresorhus** [comment]
> How do you want me to add a test? Could you please elaborate?

Some tests that prove the new functionality works correctly by executing the binary with various input. For example, create a very deep nested path and try to delete it. Create a path with Unicode characters and try to delete it. Could use a Batch script for this, unless you can think of a better way? Example: https://github.com/sindresorhus/win-clipboard/blob/218dcffb9ee5b4bed47bad42b0f1cde654b35781/test.bat

**Comment by sindresorhus** [comment]
> IShellItem can be created with long paths using simple pidls (SHSimpleIDListFromPath), 
> \- https://www.zabkat.com/blog/max-path-programmers-cookbook.htm

**Comment by medusalix** [comment]
Yeah, I found that function too, but if take a look at the [docs](https://docs.microsoft.com/en-us/windows/desktop/api/shobjidl_core/nf-shobjidl_core-shsimpleidlistfrompath) you can see that it's marked as deprecated.

**Comment by medusalix** [comment]
So I justed tried the other shell functions we found (SHSimpleIDListFromPath, SHCreateItemFromParsingName, ILCreateFromPath) and none of them works with a prefix of `\\?\` 😢. The one that seems to support the longest paths is `ILCreateFromPath`.

**Comment by medusalix** [comment]
@sindresorhus I've added some test cases. Please take a look at the results:

|            | SHFileOperation                                                              | IFileOperation                                                               |
|------------|------------------------------------------------------------------------------|------------------------------------------------------------------------------|
| Windows 7  | `Test case #1: FAILED`<br> `Test case #2: FAILED`<br> `Test case #3: FAILED` | `Test case #1: PASSED`<br> `Test case #2: PASSED`<br> `Test case #3: FAILED` |
| Windows 10 | `Test case #1: PASSED`<br> `Test case #2: PASSED`<br> `Test case #3: PASSED` | `Test case #1: PASSED`<br> `Test case #2: PASSED`<br> `Test case #3: PASSED` |

**Comment by sindresorhus** [comment]
If I run the binary from the `cmd.exe` on Windows 10, it interactively prompts me if I'm sure I want to trash the item. It should not do that.

screen shot 2019-02-11 at 17 38 54

**Comment by medusalix** [comment]
@sindresorhus That's a good catch! Apparently, it's possible to suppress that warning by implementing a [custom subclass](http://forums.codeguru.com/showthread.php?545251-How-to-tell-if-a-file-or-a-folder-could-be-placed-into-the-Recycle-Bin&p=2155389#post2155389) of the `IFileOperationProgressSink` interface. The only other alternative would be an [undocumented internal Windows API](https://stackoverflow.com/a/23721071).
You can take a look at the code if you want, but be warned, there's an insane amount of boilerplate needed to resolve that single problem, because COM requires you to implement **EVERY SINGLE FUNCTION** of that interface.

**Comment by sindresorhus** [comment]
Can you move the interface implementation into a separate file?

**Comment by medusalix** [comment]
@sindresorhus Sure, the code looks much nicer now 😊. I would really to add a test case for this scenario, but I'm out of ideas on how to do that (would need some kind of fake USB/Network Drive).

**Comment by robertguetzkow** [comment]
I think there is a problem with the code depending on how you intend to compile it. Some flags like `FOFX_ADDUNDORECORD` were added in Windows 8. Therefore this won't compile on Windows 7 even when using the version check `IsWindows8OrGreater()` because the flag isn't defined.

**Comment by medusalix** [comment]
@robertguetzkow That's correct, do you think we should add an `#ifdef` to fall back to the old flags on Windows 7?

We definitely need to compile the release on Windows 8 or 10 to be able to support the new behavior, for use in [other projects](https://github.com/sindresorhus/trash/tree/master/lib).

**Comment by sindresorhus** [comment]
I don’t think we should add if-defs. I’m fine with it only being possible to compile it on Windows 10.

**Comment by robertguetzkow** [comment]
@medusalix @sindresorhus You could add the following to define the flags that aren't available in the headers of Windows Vista and 7.
```
#ifndef FOFX_ADDUNDORECORD
#define FOFX_ADDUNDORECORD 0x20000000
#endif

#ifndef FOFX_RECYCLEONDELETE
#define FOFX_RECYCLEONDELETE 0x00080000
#endif
```

File: build.bat [hunk 4]
```diff
@@ -2,4 +2,4 @@
 
 set filename=recycle-bin
 
-gcc "%filename%".c -municode -O2 -s -o "%filename%".exe -std=c99
+gcc "%filename%".c -municode -O2 -lole32 -luuid -s -o "%filename%".exe -std=c99
```
Comment (sindresorhus) [inline build.bat]: Let's target the latest C version:

```suggestion
gcc "%filename%".c -municode -O2 -lole32 -luuid -s -o "%filename%".exe -std=c17
```

**Comment by sindresorhus** [comment]
@medusalix Anything left before this can be merged?

**Comment by medusalix** [comment]
@sindresorhus Ready to merge 😃.

**Comment by medusalix** [comment]
@sindresorhus Just a little reminder that this can be merged 👍.

**Comment by sindresorhus** [comment]
Sorry for the delay. I had to get my Windows VM fixed so I could test it.

It tried compiling a binary, but it's not really working. I have a file named `foo` on my Windows VM desktop, and when I run `recycle-bin.exe foo` nothing happens. No error message. Not moved to the trash.

**Comment by alexeyneu** [comment]
>IFileOperation can only be applied in a single-threaded apartment (STA) situation. It cannot be used >for a multithreaded apartment (MTA) situation. For MTA, you still must use SHFileOperation.  
  
https://docs.microsoft.com/en-us/windows/desktop/api/shobjidl_core/nn-shobjidl_core-ifileoperation  
last of the page  
  
and google what means `COINIT_MULTITHREADED`  
https://github.com/medusalix/recycle-bin/blob/34a86c5e8c1d1ddf8d28963b9acb52690a91e446/recycle-bin.c#L39

**Comment by medusalix** [comment]
@alexeyneu Good catch 😄. I always choose `COINIT_MULTITHREADED`, because I'm not using the Windows message loop:

> If you specify apartment threading, you are making the following guarantees:
> 
> - You will access each COM object from a single thread; you will not share COM interface pointers between multiple threads.
> - The thread will have a **message loop**. (See Window Messages in Module 1.)

But apparently it works 👍.

**Comment by sindresorhus** [comment]
Great. Works now :)

**Comment by medusalix** [comment]
Awesome 😄.

**Pull request merged**